x-pack/legacy/plugins/ml/public/application/util/url_state.ts

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { useCallback } from 'react';
+import { isEqual } from 'lodash';
+// @ts-ignore
+import queryString from 'query-string';
+import { decode, encode } from 'rison-node';
+import { useHistory, useLocation } from 'react-router-dom';
+
+import { Dictionary } from '../../../common/types/common';
+
+import { getNestedProperty } from './object_utils';
+
+export type SetUrlState = (attribute: string | Dictionary<any>, value?: any) => void;
+export type UrlState = [Dictionary<any>, SetUrlState];
+
+function getUrlState(search: string) {
+  const urlState: Dictionary<any> = {};
+  const parsedQueryString = queryString.parse(search);
+
+  try {
+    Object.keys(parsedQueryString).forEach(a => {
+      urlState[a] = decode(parsedQueryString[a]) as Dictionary<any>;
+    });
+  } catch (error) {
+    // eslint-disable-next-line no-console
+    console.error('Could not read url state', error);
+  }
+
+  return urlState;
+}
+
+// Compared to the original appState/globalState,
+// this no longer makes use of fetch/save methods.
+// - Reading from `location.search` is the successor of `fetch`.
+// - `history.push()` is the successor of `save`.
+// - The exposed state and set call make use of the above and make sure that
+//   different urlStates(e.g. `_a` / `_g`) don't overwrite each other.
+export const useUrlState = (accessor: string): UrlState => {
+  const history = useHistory();
+  const { search } = useLocation();
+
+  const setUrlState = useCallback(
+    (attribute: string | Dictionary<any>, value?: any) => {
+      const urlState = getUrlState(search);
+      const parsedQueryString = queryString.parse(search);
+
+      if (!Object.prototype.hasOwnProperty.call(urlState, accessor)) {
+        urlState[accessor] = {};
+      }
+
+      if (typeof attribute === 'string') {
+        if (isEqual(getNestedProperty(urlState, `${accessor}.${attribute}`), value)) {
```

Comment:
```suggestion
        if ((urlState?.accessor?.attribute === value)) {
```

Reply:
we cannot change it like that unfortunately. `accessor` and `attribute` are not hard-coded but are used as variables to check against something like `urlState?._g?.jobIds`. `getNestedProperty` is a TypeScript aware version similar to lodash's `get`.